### ISSUE TITLE
Value of non-Tensor arguments is ignored in torch.compile

### ISSUE THREAD

**Issue by AShedko** [pytorch/pytorch#117496]
### 🐛 Describe the bug

When compiling a method or a free function with a float (or int) parameter, the value of the parameter is ignored after the first call of the function and the value at first call is used.
```py
from __future__ import annotations

import torch
from torch import Tensor

class F():

    def __init__(self, V, r):
        self.t = V
        self.r = r

    def b(self):
        self.t = self.t * self.r
        print(f"t={self.t}")

    @torch.compile
    def method(
        self,
        A: Tensor,
        B: Tensor,
        t: float = 1.0
    ) -> Tensor:
        print(f"method: T inside is {t}, self.t is {self.t}")
        return A + B + t

@torch.compile
def free_fun(
    A: Tensor,
    B: Tensor,
    t: float = 1.0
) -> Tensor:
    print(f"free_fun: T inside is {t}")
    return A + B + t

torch.set_default_device('cuda')
g = F(16,0.5)
Z = torch.ones(3,3,1)
for _ in range(4):
    g.b()
    g.method(Z, Z, t = g.t)
    free_fun(Z, Z, g.t)
    print('----')
```

The output is:
```shell
python3 example.py 
t=8.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=4.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=2.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=1.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
```
Notice that `self.t` is also stuck as a snapshot of the object `g` at first call.
Without `@torch.compile` the output is:
```shell
t=8.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=4.0
method: T inside is 4.0, self.t is 4.0
free_fun: T inside is 4.0
----
t=2.0
method: T inside is 2.0, self.t is 2.0
free_fun: T inside is 2.0
----
t=1.0
method: T inside is 1.0, self.t is 1.0
free_fun: T inside is 1.0
----
```
Floats wrapped into scalar tensors do not have this problem.

### Versions

--2024-01-15 14:20:31--  https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 22068 (22K) [text/plain]
Saving to: ‘collect_env.py’

collect_env.py                         100%[==========================================================================>]  21.55K  --.-KB/s    in 0s      

2024-01-15 14:20:31 (53.7 MB/s) - ‘collect_env.py’ saved [22068/22068]

Collecting environment information...
PyTorch version: 2.1.2+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-1ubuntu1~22.04) 9.5.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.2.0-36-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000

Nvidia driver version: 525.147.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper PRO 5955WX 16-Cores
CPU family:                         25
Model:                              8
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        7031.2500
CPU min MHz:                        1800.0000
BogoMIPS:                           7984.87
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          512 KiB (16 instances)
L1i cache:                          512 KiB (16 instances)
L2 cache:                           8 MiB (16 instances)
L3 cache:                           64 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] fast-pytorch-kmeans==0.2.0.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.2
[pip3] pytorch-lightning==2.1.2
[pip3] pytorch-memlab==0.3.0
[pip3] torch==2.1.2+cu118
[pip3] torchaudio==2.1.2+cu118
[pip3] torchmetrics==1.2.1
[pip3] torchvision==0.16.2+cu118
[pip3] triton==2.1.0
[conda] fast-pytorch-kmeans       0.2.0.1                  pypi_0    pypi
[conda] numpy                     1.26.2          py310hb13e2d6_0    conda-forge
[conda] pytorch-lightning         2.1.2                    pypi_0    pypi
[conda] pytorch-memlab            0.3.0                    pypi_0    pypi
[conda] torch                     2.1.2+cu118              pypi_0    pypi
[conda] torchaudio                2.1.2+cu118              pypi_0    pypi
[conda] torchmetrics              1.2.1                    pypi_0    pypi
[conda] torchvision               0.16.2+cu118             pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by ezyang** [comment]
This is the same thing as https://github.com/pytorch/pytorch/issues/100061

cc @anijain2305 @colesbury

**Comment by anijain2305** [comment]
@AShedko I am unable to repro this with the latest `main` branch. I see same results w/ or w/o  torch.compile.

w/ torch.compile

```
t=8.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=4.0
method: T inside is 4.0, self.t is 4.0
free_fun: T inside is 4.0
----
t=2.0
method: T inside is 2.0, self.t is 2.0
free_fun: T inside is 2.0
----
t=1.0
method: T inside is 1.0, self.t is 1.0
free_fun: T inside is 1.0
----
```

w/o torch.compile

```
t=8.0
method: T inside is 8.0, self.t is 8.0
free_fun: T inside is 8.0
----
t=4.0
method: T inside is 4.0, self.t is 4.0
free_fun: T inside is 4.0
----
t=2.0
method: T inside is 2.0, self.t is 2.0
free_fun: T inside is 2.0
----
t=1.0
method: T inside is 1.0, self.t is 1.0
free_fun: T inside is 1.0
----
```

Can you please re-run and check?

**Comment by AShedko** [comment]
I tried it on `nightly` and could not reproduce it either.
So it was fixed somewhere between `nightly` and `2.1.2`.

**Comment by anijain2305** [comment]
Yeah, seems like it. Closing the issue for now.
